src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -72,50 +73,75 @@ internal int FindMetricAggregators(ReadOnlySpan<KeyValuePair<string, object>> ta
 
             int aggregatorIndex;
 
-            lock (this.lockKeyValue2MetricAggs)
+            string[] seqKey = null;
+
+            // GetOrAdd by TagKey at 1st Level of 2-level dictionary structure.
+            // Get back a Dictionary of [ Values x Metrics[] ].
+            if (!this.keyValue2MetricAggs.TryGetValue(tagKey, out var value2metrics))
             {
-                string[] seqKey = null;
+                // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
+                seqKey = new string[len];
+                tagKey.CopyTo(seqKey, 0);
 
-                // GetOrAdd by TagKey at 1st Level of 2-level dictionary structure.
-                // Get back a Dictionary of [ Values x Metrics[] ].
-                if (!this.keyValue2MetricAggs.TryGetValue(tagKey, out var value2metrics))
+                value2metrics = new ConcurrentDictionary<object[], int>(new ObjectArrayEqualityComparer());
```

Comment:
Could we promote `new ObjectArrayEqualityComparer()` to a static and then reuse it on all the inner dictionaries?